Update unix.mli

craff opened the pull request:
address #12330

craff committed:
Update unix.mli

address #12330

craff committed:
Update unix.mli

spelling

craff commented:
I wnated to fix the hygiene pb: 
`This should be fixable by just running tools/sync_stdlib_docs and reviewing the changes it makes.`

But  I can't find `tools/sync_stdlib_docs` ?

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -227,6 +227,14 @@ val execv : string -> string array -> 'a
    Using {!create_process} or one of the [open_process_*] functions
    instead is recommended.
 
+   The first element of the array [args] will often be [argv[0]] or
```
Using `will often be` doesn't really make sense for a specification, you are not documenting an optional behavior: programs will break if they are not send the correct number of arguments.

Octachron replied:
An alternative wording might be:

> On both unix and windows, the convention is that the first element of the array [args] is the name of the executed program.

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -227,6 +227,14 @@ val execv : string -> string array -> 'a
 
+   The first element of the array [args] will often be [argv[0]] or
+   [Sys.argv.(0)] in the new process [prog]. This is plateform
+   dependant and you should refer to the man page of execv, which for
```
`man page` are already  Unix-only.

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -865,7 +873,9 @@ val create_process :
    The executable file [prog] is searched in the path.
-   The new process has the same environment as the current process. *)
+   The new process has the same environment as the current process. 
+   The first element of the array [args] may or may not be required to
```
`may or may not be required` sounds mostly confusing.

craff replied:
I was vague because of your comments on the same issue... I will rephrase as you suggest, but do you have a system in mind that does not follow the convention of first args being the process name ?

Maybe what is not clear is what happens when the first args is not the same as the process name.

In fact, my main concern is that if you do

Unix.open_process_args "cmd" [| "-l" |] it will not work as expected while 
Unix.open_process_args "cmd" [| ""; "-l" |] probably will (I did not test), because most commands look for option from argv[1] ?

Octachron replied:
What I have in mind is that the specification of `program name` may vary. Typically, the C17 standard reads as
```
If the value of argc is greater than zero, the string pointed to by argv[0]
represents the program name; argv[0][0] shall be the null character if the
program name is not available from the host environment.
```
but the `represent` verb is quite ambiguous.

Famously, the busybox executable mimics the behavior of the program which name in its `argv[0]`. 

Overall, we should focus on the important points: we want to warn users that `argv[0]` has a special meaning.

craff committed:
Update unix.mli

- rephrase by copying the doc of create_process which was ok to execve and open_process

- also clarify process_pid and alike

craff commented:
@Octachron : I rephrased the PR and also applied your suggestion of documentation in issue #12329.

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -217,7 +217,9 @@ type wait_flag =
 (** [execv prog args] execute the program in file [prog], with
-   the arguments [args], and the current process environment.
+   the arguments [args], and the current process environment.  
+   Note that the first argument is by convention the filename of
```
Here, "first argument" feels slightly ambiguous to me: it may refer to `prog`. I would propose to be very explicit: "the first argument, [args.(0)], "

Octachron commented:
The `sync`ing tool is in the `tools` directory at the top of the tree, but you should transfer your change to `unixLabels.mli` by hand first (otherwise the tools will erase your changes).

craff committed:
clarification of the documentation of process related function un unixLabels.mli

craff committed:
Changes

craff commented:
> The `sync`ing tool is in the `tools` directory at the top of the tree, but you should transfer your change to `unixLabels.mli` by hand first (otherwise the tools will erase your changes.

And you need to compile ocaml to have it ! I got it, thanks.

craff commented:
I do not understant why one test in the test suite failed while I only changed comments ?

lthls commented:
> I do not understant why one test in the test suite failed while I only changed comments ?

It is a known, unrelated issue (see #12345 for a few more details).

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -216,10 +216,11 @@ type wait_flag =
+(** [execv prog args] execute the program in file [prog], with the arguments
+   [args], and the current process environment.  Note that the first
+   argument, [args.(0)], is by convention the filename of the program being
+   executed, just like [Sys.argv.(0)] These [execv*] functions never return:
```
Missing dot: `like [Sys.argv.(0)] These [execv*]`  ⇒ `just like [Sys.argv.(0)]. These [execv*]`

Octachron commented on otherlibs/unix/unix.mli:
```diff
@@ -854,18 +855,19 @@ val mkfifo : string -> file_perm -> unit
+    program being executed, just like [Sys.argv.(0)].
+
+    The pid of the new process is returned immediately; the new process
+    executes concurrently with the current process.  The standard input and
```
If you are separating paragraphs by a newline, you should do so here too (and in the rest of the paragraph below).

Octachron commented on Changes:
```diff
@@ -97,6 +97,10 @@ Working version
 
+- #12338: clarification of the documentation of process related function in
+  the unix module regarding the first element of args and shell's pid.
+  (Christophe Raffalli review by Florian Angeletti)
```
missing `,` before review.

craff committed:
apply suggestions

Octachron reviewed (approved):
I agree that documenting those potential pitfalls is worthwhile (and not only for beginners). Thanks !

Octachron commented:
Note that the the documentation needs to be synced, and the Changes entry updated.

craff committed:
sync doc

craff committed:
Merge branch 'trunk' into patch-1

craff commented:
Actually the "Resolve Conflict" button in github does a better job that git on my laptop ;-(

craff committed:
trailing white space in Changes
